modules/phase_field/src/utils/PolycrystalICTools.C

```diff
@@ -83,3 +92,256 @@ PolycrystalICTools::assignPointToGrain(const Point & p, const std::vector<Point>
+    {
+      const Elem * neighbor_ancestor = elem->neighbor(i);
+      if (neighbor_ancestor)
+        // Retrieve only the active neighbors for each side of this element, append them to the list of active neighbors
+        neighbor_ancestor->active_family_tree_by_neighbor(all_active_neighbors, elem, false);
+    }
+
+    // Loop over all active element neighbors
+    for (std::vector<const Elem *>::const_iterator neighbor_it = all_active_neighbors.begin();
+         neighbor_it != all_active_neighbors.end(); ++neighbor_it)
+    {
+      const Elem * neighbor = *neighbor_it;
+      std::map<dof_id_type, unsigned int>::const_iterator grain_it2 = element_to_grain.find(neighbor->id());
+      mooseAssert(grain_it2 != element_to_grain.end(), "Element not found in map");
+      unsigned int their_grain = grain_it2->second;
+
+      if (my_grain != their_grain)
+      {
+        /**
+         * We've found a grain neighbor interface. In order to assign order parameters though, we need to make
+         * sure that we build out a small buffer region to avoid literal "corner cases" where nodes on opposite
+         * corners of a QUAD end up with the same OP because those nodes are not nodal neighbors.
+         *
+         * To do that we'll build a halo region based on these interface nodes. For now, we need to record
+         * the nodes inside of the grain and those outside of the grain.
+         */
+
+        // First add corresponding element and grain information
+        local_ids[my_grain].insert(elem->id());
+        local_ids[their_grain].insert(neighbor->id());
+
+        // Now add opposing element and grain information
+        halo_ids[my_grain].insert(neighbor->id());
+        halo_ids[their_grain].insert(elem->id());
+      }
+      //  adjacency_matrix[my_grain][their_grain] = 1;
+    }
+  }
+
+  // Build up the halos
+  std::set<dof_id_type> set_difference;
+  for (unsigned int i = 0; i < n_grains; ++i)
+  {
+    std::set<dof_id_type> orig_halo_ids(halo_ids[i]);
+
+    for (unsigned int halo_level = 0; halo_level < 2; ++halo_level)
+    {
+      for (std::set<dof_id_type>::iterator entity_it = orig_halo_ids.begin();
+           entity_it != orig_halo_ids.end(); ++entity_it)
+      {
+        if (true)
+          visitElementalNeighbors(mesh.elemPtr(*entity_it), halo_ids[i]);
+        else
+          mooseError("Unimplemented");
+      }
+
+      set_difference.clear();
+      std::set_difference(halo_ids[i].begin(), halo_ids[i].end(), local_ids[i].begin(), local_ids[i].end(),
+                          std::insert_iterator<std::set<dof_id_type> >(set_difference, set_difference.begin()));
+
+      halo_ids[i].swap(set_difference);
+    }
+  }
+
+  // Finally look at the halo intersections to build the connectivity graph
+  std::set<dof_id_type> set_intersection;
+  for (unsigned int i = 0; i < n_grains; ++i)
+    for (unsigned int j = i+1; j < n_grains; ++j)
+    {
+      if (i == j)
```

Comment:
You guys are both hilarious. The first version of this loop had `j = 0` ... I later changed it since the adjacency graph is always symmetrical. I'll remove the conditional.